src/main/java/net/openhft/hashing/LongTupleHashFunction.java

```diff
@@ -0,0 +1,779 @@
+package net.openhft.hashing;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import javax.annotation.ParametersAreNonnullByDefault;
+import sun.nio.ch.DirectBuffer;
+
+import java.io.Serializable;
+import java.nio.ByteBuffer;
+
+import static net.openhft.hashing.CharSequenceAccess.nativeCharSequenceAccess;
+import static net.openhft.hashing.UnsafeAccess.*;
+import static net.openhft.hashing.Util.*;
+
+/**
+ * Tuple hash function producing {@code long[]} result from byte sequences of any length and
+ * a plenty of different sources which "feels like byte sequences". Except {@link
+ * #hashBytes(byte[])}, {@link #hashBytes(ByteBuffer)} (with their "sliced" versions) and
+ * {@link #hashMemory(long, long)} methods, which actually accept byte sequences, notion of byte
+ * sequence is defined as follows:
+ * <ul>
+ *     <li>For methods accepting arrays of Java primitives, {@code String}s and
+ *     {@code StringBuilder}s, byte sequence is how the input's bytes are actually lay in memory.
+ *     </li>
+ *     <li>For methods accepting single primitive values, byte sequence is how this primitive
+ *     would be put into memory with {@link ByteOrder#nativeOrder() native} byte order, or
+ *     equivalently, {@code hashXxx(primitive)} has always the same result as {@code
+ *     hashXxxs(new xxx[] {primitive})}, where "xxx" is any Java primitive type name.</li>
+ *     <li>For {@link #hash(Object, Access, long, long)} method byte sequence abstraction
+ *     is defined by the given {@link Access} strategy to the given object.</li>
+ * </ul>
+ *
+ * <p>Tuple hash function implementation could either produce equal results for equal input on platforms
+ * with different {@link ByteOrder}, favoring one byte order in terms of performance, or different
+ * results, but performing equally good. This choice should be explicitly documented for all
+ * {@code LongTupleHashFunction} implementations.
+ *
+ * Each hash api have two forms:
+ * <ul>
+ *     <li> {code void hash(input..., long[] result)} will put the hash result in array {@code result}
```

Comment:
`@code`

Reply:
Should also specify what will happen if the provided `result` array is shorter than needed, longer than needed, and `null` (throw NPE)